v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -0,0 +1,455 @@
+            # Thus, we know that the state will not be changing on us within this block.
+            if not self.is_connected():
+                # Record the type of operation in progress
+                self._connection_lock.connection_type = OP_TYPE_CONNECT
+
+                # Start listening for connect before performing Paho connect to make sure
+                # we don't miss it. This could possibly happen due to timing issues on the
+                # Paho thread (i.e. CONNACK received after invoking connect, but before our
+                # listening tasks have started)
+                async def wait_for_success():
+                    async with self.connected_cond:
+                        await self.connected_cond.wait()
+
+                async def wait_for_failure():
+                    async with self._connect_failed_cond:
+                        await self._connect_failed_cond.wait()
+
+                success = asyncio.create_task(wait_for_success())
+                failure = asyncio.create_task(wait_for_failure())
+
+                # Start the reconnect daemon (if enabled and not already running)
+                if self._auto_reconnect and not self._reconnect_daemon:
+                    self._reconnect_daemon = asyncio.create_task(self._reconnect_loop())
+
+                # Make sure the tasks are running
+                await asyncio.sleep(0)
+
+                # Paho Connect
+                logger.debug("Attempting connect using port {}".format(self._port))
+                try:
+                    rc = await self._loop.run_in_executor(
+                        None,
+                        functools.partial(
+                            self._mqtt_client.connect,
+                            host=self._hostname,
+                            port=self._port,
+                            keepalive=self._keep_alive,
+                        ),
+                    )
+                    message = mqtt.error_string(rc)
+                    logger.debug("Connect returned rc {} - {}".format(rc, message))
+                # TODO: more specialization of errors
+                except Exception as e:
+                    # Clean up listener tasks
+                    success.cancel()
+                    failure.cancel()
+                    raise exceptions.ConnectionFailedError("Connect failed") from e
+
+                if rc != mqtt.MQTT_ERR_SUCCESS:
+                    # Clean up listener tasks
+                    success.cancel()
+                    failure.cancel()
+                    raise exceptions.ConnectionFailedError(
+                        "Connect returned rc {} - {}".format(rc, message)
+                    )
+
+                # Start Paho network loop. If already started, this will return a fail code,
+                # but we don't really care - no harm, no foul.
+                logger.debug("Starting Paho network loop")
+                self._mqtt_client.loop_start()
```

Comment:
I stronly recomment asserting `self._mqtt_client._thread` and `self._mqtt_client._thread.is_alive()`. There are paths through Paho that stop the thread but don't set `_thread` to `None`, and `loop_start` only checks that the `_thread` object is not `None`. It doesn't check if the thread is actually alive.  This assert will hopefully never fire, but when it does, it will save many hours of debugging.

Reply:
My understanding is that Paho only can't null the thread if loop_stop() is called from within the loop itself, and I've explicitly avoided doing that as a result. Are you saying there's another situation where that can happen?

Reply:
Any time `loop_forever` exits for any reason besides `loop_stop` being called. This is why you have to call `loop_stop` when there's an unexpected disconnect. Maybe it would be more robust to check `_thread.is_alive()` here and null the thread before calling `loop_start`. This way you don't have to rely on `on_disconnect` being called. This can happen if there is an exception inside `loop_forever`. Now that I think about it, we probably have some "client is bricked" bugs in v2 that could be fixed this way.

Reply:
Added a check to the rc of loop_start